core/src/main/java/org/infinispan/xsite/spi/XSiteConflictResolver.java

```diff
@@ -0,0 +1,41 @@
+package org.infinispan.xsite.spi;
+
+import java.util.concurrent.CompletionStage;
+
+import org.infinispan.metadata.Metadata;
+
+/**
+ * An interface to resolve conflicts for asynchronous cross-site replication.
+ *
+ * @author Pedro Ruivo
+ * @since 12.0
+ */
+public interface XSiteConflictResolver<K, V> {
+
+   /**
+    * Resolves conflicts for asynchronous cross-site replication.
+    * <p>
+    * When a conflict is detected (concurrent updates on the same key in different sites), this method is invoked with
+    * the local data and the remote site's data ({@link SiteEntry}). It includes the value and the {@link Metadata}
+    * associated.
+    * <p>
+    * The value and the {@link Metadata} may be {@code null}. If that is the case, it means the {@code key} doesn't
+    * exist (for {@code localEntry}) or it is a remove operation (for {@code remoteEntry}).
+    * <p>
+    * The returned {@link SiteEntry} must be equal independent of the order of the arguments (i.e. {@code resolve(k, s1,
+    * s2).equals(resolve(k, s2, s1))}) otherwise your date may be corrupted. It is allowed to return one of the
+    * arguments ({@code localEntry} or {@code remoteEntry}) and to create a new {@link SiteEntry} with a new value.
+    * <p>
+    * Note: if the return {@link SiteEntry#getValue()} is {@code null}, Infinispan will interpret it to remove the
+    * {@code key}.
+    * <p>
+    * Note2: This method shouldn't block (I/O or locks). If it needs to block, use a different thread and complete the
+    * {@link CompletionStage} with the result.
```

Comment:
We should probably propose that users utilise the `BlockingManager` here.

Reply:
done